src/app/learnocaml_common.ml

```diff
@@ -969,3 +969,271 @@ let setup_prelude_pane ace prelude =
+      display_link (display_skill_meta s index) content_id skill
+      
+    let display_exercise_link content_id meta e =
+      display_link (display_exercise_meta e meta) content_id e
+      
+    let display_authors authors =
+      let open Tyxml_js.Html5 in
+      let author (name, mail) =
+        span [ pcdata name;
+               pcdata " <";
+               a ~a:[ a_href ("mailto:" ^ mail) ]
+                 [ pcdata mail ];
+               pcdata ">"
+          ] in
+      display_list @@ List.map author authors
+
+    let display_meta token ex_meta id =
+      let open Learnocaml_data.Exercise in
+      let open Tyxml_js.Html5 in
+      let ident =
+        Format.asprintf "%s %s" [%i "Identifier:" ] id in
+      let opt l f = match l with [] -> None | l -> Some (f l) in
+      let authors =
+        opt ex_meta.Meta.author @@ fun author ->
+                                   let title = match author with [_] -> [%i"Author"] | _ -> [%i"Authors"] in
+                                   span [ pcdata title; pcdata " " ] ::
+                                     display_authors author
+      in
+      retrieve (Learnocaml_api.Exercise_index token)
+      >|= fun (index, _) ->
+      let req_map, focus_map =
+        Exercise.Index.fold_exercises (fun (req, focus) id meta ->
+            let add sk id map =
+              SMap.add sk
+                (SSet.add id (try SMap.find sk map with Not_found -> SSet.empty))
+                map
+            in
+            List.fold_left (fun acc sk -> add sk id acc) req
+              meta.Exercise.Meta.requirements,
+            List.fold_left (fun acc sk -> add sk id acc) focus
+              meta.Exercise.Meta.focus
+          ) (SMap.empty, SMap.empty) index
+      in
+      let focus =
+        opt ex_meta.Meta.focus @@ fun focus ->
+                                  [%i "Skills trained:"],
+                                  display_list ~sep:(pcdata "") @@
+                                    List.map (fun s ->
+                                        display_skill_link
+                                          (try SSet.elements (SMap.find s focus_map) with Not_found -> [])
+                                          "learnocaml-exo-focus-meta" (`Focus s))
+                                      focus in
+      let requirements =
+        opt ex_meta.Meta.requirements @@ fun requirements ->
+                                         [%i "Skills required:"],
+                                         display_list ~sep:(pcdata "") @@
+                                           List.map (fun s ->
+                                               display_skill_link
+                                                 (try SSet.elements (SMap.find s req_map) with Not_found -> [])
+                                                 "learnocaml-exo-requirements-meta" (`Requirements s))
+                                             requirements in
```

Comment:
Same here: there are some factorization opportunities that are missing. The introduction of well-designed combinators should help. Please remember that a good functional programming style is made of composing small and general functions... I know that the source code of learn-ocaml has not been initially written with that kind of stylistic consideration in mind but I would like to improve on that.